mcstatus/forge_data.py

```diff
@@ -0,0 +1,260 @@
+"""
+
+from __future__ import annotations
+
+import io
+from dataclasses import dataclass
+from typing import Final, IO, TYPE_CHECKING
+
+from mcstatus.protocol.connection import BaseConnection, BaseReadSync, Connection
+
+VERSION_FLAG_IGNORE_SERVER_ONLY: Final = 0b1
+IGNORE_SERVER_ONLY: Final = "<not required for client>"
+
+
+if TYPE_CHECKING:
+    from typing_extensions import NotRequired, Self, TypedDict
+
+    class RawForgeDataChannel(TypedDict):
+        res: str
+        """Channel name and ID (for example ``fml:handshake``)."""
+        version: str
+        """Channel version (for example ``1.2.3.4``)."""
+        required: bool
+        """Is this channel required for client to join?"""
+
+    class RawForgeDataMod(TypedDict):
+        modid: NotRequired[str]
+        modId: NotRequired[str]  # noqa: N815 # camel case
+        modmarker: NotRequired[str]
+        """Mod version."""
+        version: NotRequired[str]
+
+    class RawForgeData(TypedDict):
+        fmlNetworkVersion: NotRequired[int]  # noqa: N815 # camel case
+        channels: NotRequired[list[RawForgeDataChannel]]
+        mods: NotRequired[list[RawForgeDataMod]]
+        modList: NotRequired[list[RawForgeDataMod]]  # noqa: N815 # camel case
+        d: NotRequired[str]
+        truncated: NotRequired[bool]
+
+else:
+    RawForgeDataChannel = dict
+    RawForgeDataMod = dict
+    RawForgeData = dict
+
+
+@dataclass
+class ForgeDataChannel:
+    res: str
+    """Channel name and ID (for example ``fml:handshake``)."""
+    version: str
+    """Channel version (for example ``1.2.3.4``)."""
+    required: bool
+    """Is this channel required for client to join?"""
+
+    @classmethod
+    def build(cls, raw: RawForgeDataChannel) -> Self:
+        """Build an object about Forge channel from raw response.
+
+        :param raw: ``channel`` element in raw forge response :class:`dict`.
+        :return: :class:`ForgeDataChannel` object.
+        """
+        return cls(res=raw["res"], version=raw["version"], required=raw["required"])
+
+    @classmethod
+    def decode(cls, buffer: Connection, mod_id: str | None = None) -> Self:
+        """Decode an object about Forge channel from decoded optimized buffer.
+
+        :param buffer: :class:`Connection` object from UTF-16 encoded binary data.
+        :param mod_id: Optional mod id prefix :class:`str`.
+        :return: :class:`ForgeDataChannel` object.
+        """
+        channel_identifier = buffer.read_utf()
+        if mod_id is not None:
+            channel_identifier = f"{mod_id}:{channel_identifier}"
+        version = buffer.read_utf()
+        client_required = buffer.read_bool()
+
+        return cls(
+            res=channel_identifier,
+            version=version,
+            required=client_required,
+        )
+
+
+@dataclass
+class ForgeDataMod:
+    mod_id: str
+    mod_version: str
```

Comment:
In v2, Forge renamed it from `version` to `marker`, and I think without reason. Let's follow the newer protocol and use `marker` instead of `version` (but in docstring we still can say, that it's a version almost all the time).

Reply:
I feel like it would be more intuitive from a user's perspective for the attribute to be named version. Marker is more vague and does not communicate nearly as well what that data means.

Reply:
It doesn't need to be a version, it's a marker, actually. For example, see https://github.com/py-mine/mcstatus/pull/578/files#diff-1cf00a171f2c02af35ced6962cd70cfafaf8b96189dd77422f778a11770ec370R402 or https://github.com/py-mine/mcstatus/pull/578/files#diff-1cf00a171f2c02af35ced6962cd70cfafaf8b96189dd77422f778a11770ec370R426.

Reply:
I don't quite understand why this PR is what is linked to

Reply:
Those weren't permalinks, so they got outdated. Here is newer version

https://github.com/py-mine/mcstatus/blob/76f728ba8a0e178ed88de6755ea582d9496c601b/tests/status_response/test_java.py#L380
https://github.com/py-mine/mcstatus/blob/76f728ba8a0e178ed88de6755ea582d9496c601b/tests/status_response/test_java.py#L426

Reply:
Renamed `marker` in 962e300